stratosphere/loader/source/ldr_capabilities.cpp

```diff
@@ -0,0 +1,428 @@
+
+        DEFINE_CAPABILITY_CLASS(KernelVersion,
+            u32 GetMinorVersion() const {
+                return (this->GetValue() >> 0) & 0xF;
+            }
+
+            u32 GetMajorVersion() const {
+                /* TODO: Are upper bits unused? */
+                return (this->GetValue() >> 4) & 0x1FFF;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        return restrict.GetValue() == this->GetValue();
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(HandleTable,
+            u32 GetSize() const {
+                return (this->GetValue() >> 0) & 0x3FF;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        return this->GetSize() <= restrict.GetSize();
+                    }
+                }
+                return false;
+            }
+        );
+
+        DEFINE_CAPABILITY_CLASS(DebugFlags,
+            bool GetAllowDebug() const {
+                return ((this->GetValue() >> 0) & 1) != 0;
+            }
+
+            bool GetForceDebug() const {
+                return ((this->GetValue() >> 1) & 1) != 0;
+            }
+
+            bool IsValid(const u32 *kac, size_t kac_count) const {
+                for (size_t i = 0; i < kac_count; i++) {
+                    if (GetCapabilityId(kac[i]) == this->GetId()) {
+                        const auto restrict = Decode(kac[i]);
+
+                        return (restrict.GetValue() & this->GetValue()) == this->GetValue();
+                    }
+                }
+                return false;
+            }
+
+            static constexpr u32 Encode(bool allow_debug, bool force_debug) {
+                const u32 desc = (static_cast<u32>(force_debug) << 1) | (static_cast<u32>(allow_debug) << 0);
+                return (desc << ValueShift) | IdMask;
+            }
+        );
+
+    }
+
+    /* Capabilities API. */
+    Result ValidateCapabilities(const void *acid_kac, size_t acid_kac_size, const void *aci_kac, size_t aci_kac_size) {
+        const u32 *acid_caps = reinterpret_cast<const u32 *>(acid_kac);
+        const u32 *aci_caps = reinterpret_cast<const u32 *>(aci_kac);
+        const size_t num_acid_caps = acid_kac_size / sizeof(*acid_caps);
+        const size_t num_aci_caps = aci_kac_size / sizeof(*aci_caps);
+
+        for (size_t i = 0; i < num_aci_caps; i++) {
+            const u32 cur_cap = aci_caps[i];
+            const auto id = GetCapabilityId(cur_cap);
+
+#define VALIDATE_CASE(id) \
```

Comment:
Virtual methods have overhead and probably that's not what we want to do here.

Reply:
Virtual functions just cause pointless runtime overhead, here. The generated assembly currently looks quite nice in IDA (everything gets inlined), and I'd rather not explicitly call functions.

`At the very least, the InvalidCapability##id result should be moved to be a static member on the capability class.`

Can't really do this, since we're eventually going to make sts::Result all classes (see mesosphere definition) -- I also want the result definitions all in one place (in libstrat headers).

You could do e.g. static constexpr Result InvalidResult in the class def macro, but that seems like a trivial gain? Not sure that's actually preferable style, anyway.